src/test/kotlin/com/karumi/katagenda/ui/ContactsListPresenterTest.kt

```diff
@@ -0,0 +1,163 @@
+
+        presenter.onInitialize()
+
+        verify(view)?.showEmptyCase()
+    }
+
+    @Test
+    fun shouldShowContactsFromTheAgendaOnInitialize() {
+        val presenter = givenAContactsListPresenter()
+        val contacts = givenTheAgendaIsNotEmpty()
+
+        presenter.onInitialize()
+
+        verify(view)?.showContacts(contacts)
+    }
+
+    @Test
+    fun shouldShowGoodbyeMessageOnStop() {
+        val presenter = givenAContactsListPresenter()
+
+        presenter.onStop()
+
+        verify(view)?.showGoodbyeMessage()
+    }
+
+    @Test
+    fun shouldShowTheContactsListWithTheNewContactOnContactAdded() {
+        val presenter = givenAContactsListPresenter()
+        val contactToCreate = givenTheUserAddsAContact()
+        givenTheContactIsAddedCorrectly(contactToCreate)
+
+        presenter.onInitialize()
+        presenter.onAddContactOptionSelected()
+
+        argumentCaptor<List<Contact>>().apply {
+            verify(view, times(2))?.showContacts(capture())
+
+            assertTrue(allValues[1].contains(contactToCreate))
+        }
+
+    }
+
+    @Test
+    fun shouldShowAnErrorIfTheFirstNameOfTheNewContactIsEmpty() {
+        val presenter = givenAContactsListPresenter()
+        givenTheUserTypesContactInfo("", ANY_LAST_NAME, ANY_PHONE_NUMBER)
+
+        presenter.onInitialize()
+        presenter.onAddContactOptionSelected()
+
+        verify(view)?.showDefaultError()
+    }
+
+    @Test
+    fun shouldShowAnErrorIfTheLastNameOfTheNewContactIsEmpty() {
+        val presenter = givenAContactsListPresenter()
+        givenTheUserTypesContactInfo(ANY_FIRST_NAME, "", ANY_PHONE_NUMBER)
+
+        presenter.onInitialize()
+        presenter.onAddContactOptionSelected()
+
+        verify(view)?.showDefaultError()
+    }
+
+    @Test
+    fun shouldShowAnErrorIfTheNameOfTheNewContactIsEmpty() {
+        val presenter = givenAContactsListPresenter()
+        givenTheUserTypesContactInfo(ANY_FIRST_NAME, ANY_LAST_NAME, "")
+
+        presenter.onInitialize()
+        presenter.onAddContactOptionSelected()
+
+        verify(view)?.showDefaultError()
+    }
+
+    private fun givenTheUserTypesContactInfo(t: String, anyLastName: String, anyPhoneNumber: String) {
+        whenever(view?.newContactFirstName).thenReturn(t)
+        whenever(view?.newContactLastName).thenReturn(anyLastName)
+        whenever(view?.newContactPhoneNumber).thenReturn(anyPhoneNumber)
+    }
+
+    private fun givenTheContactIsAddedCorrectly(contact: Contact) {
+        whenever(addContact!!.execute(contact)).thenReturn(contact)
+        val newContacts = ArrayList<Contact>()
+        newContacts.add(contact)
+        whenever(getContacts!!.execute()).thenReturn(newContacts)
+    }
+
+    private fun givenTheAgendaIsEmpty() {
+        whenever(getContacts!!.execute()).thenReturn(emptyList<Contact>())
+    }
+
+    private fun givenTheAgendaIsNotEmpty(): List<Contact> {
+        val contacts = ArrayList<Contact>()
+        contacts.addAll(
```

Comment:
We are creating two lists here, the one from the range and the empty array list we are filling afterwards, isn't it better (more readable) to write something like this instead?

```kotlin
val contacts = (0 to ANY_NUMBER_OF_CONTACTS - 1)
                .toList()
                .map {
                    Contact(ANY_FIRST_NAME, ANY_LAST_NAME, ANY_PHONE_NUMBER)
                }
```